drv/lpc55-swd/src/main.rs

```diff
@@ -972,6 +1013,73 @@ impl ServerImpl {
     fn pin_setup(&mut self) {
         setup_pins(self.gpio).unwrap_lite();
     }
+
+    /// Swaps the currently-active SP slot
+    fn swap_sp_slot(&mut self) -> Result<(), Ack> {
+        // All registers and constants are within the FLASH peripheral block, so
+        // I'm going to skip prefixing everything with `FLASH_`.
+
+        // RM0433 Table 8
+        const BASE: u32 = 0x52002000;
+
+        // RM0433 Section 4
+        const OPTCR: u32 = BASE + 0x018;
+        const OPTCR_OPTLOCK_BIT: u32 = 1 << 0;
+        const OPTCR_OPTSTART_BIT: u32 = 1 << 1;
+
+        // Check whether we have to unlock the flash control register
+        let optcr = self.read_single_target_addr(OPTCR)?;
+        if optcr & OPTCR_OPTLOCK_BIT != 0 {
+            // Keys constants are defined in RM0433 Rev 7
+            // Section 4.9.3
+            const OPT_KEY1: u32 = 0x0819_2A3B;
+            const OPT_KEY2: u32 = 0x4C5D_6E7F;
+            const OPTKEYR: u32 = BASE + 0x008;
+            self.write_single_target_addr(OPTKEYR, OPT_KEY1)?;
+            self.write_single_target_addr(OPTKEYR, OPT_KEY2)?;
+        }
+
+        // Read the current bank swap bit
+        const OPTSR_CUR: u32 = BASE + 0x01C;
+        const OPTSR_SWAP_BANK_OPT_BIT: u32 = 1 << 31;
+        const OPTSR_OPT_BUSY_BIT: u32 = 1 << 0;
+        let optsr_cur = self.read_single_target_addr(OPTSR_CUR)?;
+
+        // Mask and toggle the bank swap bit
+        let new_swap =
+            (optsr_cur & OPTSR_SWAP_BANK_OPT_BIT) ^ OPTSR_SWAP_BANK_OPT_BIT;
+
+        // Modify the bank swap bit in OPTSR_PRG
+        const OPTSR_PRG: u32 = BASE + 0x020;
+        let mut optsr_prg = self.read_single_target_addr(OPTSR_PRG)?;
+        optsr_prg = (optsr_prg & !OPTSR_SWAP_BANK_OPT_BIT) | new_swap;
+        self.write_single_target_addr(OPTSR_PRG, optsr_prg)?;
+
+        // Start programming option bits
+        let mut optcr = self.read_single_target_addr(OPTCR)?;
+        optcr |= OPTCR_OPTSTART_BIT;
+        self.write_single_target_addr(OPTCR, optcr)?;
+
+        // Wait for option bit programming to finish
+        loop {
+            let optsr_cur = self.read_single_target_addr(OPTSR_CUR)?;
+            if optsr_cur & OPTSR_OPT_BUSY_BIT == 0 {
+                break;
+            }
+        }
```

Comment:
nit, take it or leave it: why not
```suggestion
        while self.read_single_target_addr(OPTSR_CUR)? & OPTSR_OPT_BUSY_BIT != 0 {}
```

Reply:
Sure, done in 0526b9574a5bd51fa042459b1a5adad6a644afa5 (along with a delay)